core/extension/gdnative_interface.h

```diff
@@ -0,0 +1,416 @@
+	NATIVE_VARIANT_OP_XOR,
+	NATIVE_VARIANT_OP_NOT,
+	//containment
+	NATIVE_VARIANT_OP_IN,
+	NATIVE_VARIANT_OP_MAX
+
+};
+
+typedef void *GDNativeVariantPtr;
+typedef void *GDNativeStringNamePtr;
+typedef void *GDNativeStringPtr;
+typedef void *GDNativeObjectPtr;
+typedef void *GDNativeTypePtr;
+typedef void *GDNativeMethodBindPtr;
+typedef int64_t GDNativeInt;
+typedef uint32_t GDNativeBool;
+typedef uint64_t GDObjectInstanceID;
+
+/* VARIANT DATA I/O */
+
+enum GDNativeCallErrorType {
+	NATIVE_CALL_OK,
+	NATIVE_CALL_ERROR_INVALID_METHOD,
+	NATIVE_CALL_ERROR_INVALID_ARGUMENT, // expected is variant type
+	NATIVE_CALL_ERROR_TOO_MANY_ARGUMENTS, // expected is number of arguments
+	NATIVE_CALL_ERROR_TOO_FEW_ARGUMENTS, // expected is number of arguments
+	NATIVE_CALL_ERROR_INSTANCE_IS_NULL,
+
+};
+
+struct GDNativeCallError {
+	GDNativeCallErrorType error;
+	int32_t argument;
+	int32_t expected;
+};
+
+typedef void (*GDNativeVariantFromTypeConstructorFunc)(GDNativeVariantPtr, GDNativeTypePtr);
+typedef void (*GDNativeTypeFromVariantConstructorFunc)(GDNativeTypePtr, GDNativeVariantPtr);
+typedef void (*GDNativePtrOperatorEvaluator)(const GDNativeTypePtr p_left, const GDNativeTypePtr p_right, GDNativeTypePtr r_result);
+typedef void (*GDNativePtrBuiltInMethod)(GDNativeTypePtr p_base, const GDNativeTypePtr *p_args, GDNativeTypePtr r_return, int p_argument_count);
+typedef void (*GDNativePtrConstructor)(GDNativeTypePtr p_base, const GDNativeTypePtr *p_args);
+typedef void (*GDNativePtrSetter)(GDNativeTypePtr p_base, const GDNativeTypePtr p_value);
+typedef void (*GDNativePtrGetter)(const GDNativeTypePtr p_base, GDNativeTypePtr r_value);
+typedef void (*GDNativePtrIndexedSetter)(GDNativeTypePtr p_base, GDNativeInt p_index, const GDNativeTypePtr p_value);
+typedef void (*GDNativePtrIndexedGetter)(const GDNativeTypePtr p_base, GDNativeInt p_index, GDNativeTypePtr r_value);
+typedef void (*GDNativePtrKeyedSetter)(GDNativeTypePtr p_base, const GDNativeTypePtr p_key, const GDNativeTypePtr p_value);
+typedef void (*GDNativePtrKeyedGetter)(const GDNativeTypePtr p_base, const GDNativeTypePtr p_key, GDNativeTypePtr r_value);
+typedef bool (*GDNativePtrKeyedChecker)(const GDNativeVariantPtr p_base, const GDNativeVariantPtr p_key);
+typedef void (*GDNativePtrUtilityFunction)(GDNativeTypePtr r_return, const GDNativeTypePtr *p_arguments, int p_argument_count);
+
+typedef GDNativeObjectPtr (*GDNativeClassConstructor)();
+
+/* EXTENSION CLASSES */
+
+typedef void *GDExtensionClassInstancePtr;
+
+typedef GDNativeBool (*GDNativeExtensionClassSet)(GDExtensionClassInstancePtr p_instance, const GDNativeStringNamePtr p_name, const GDNativeVariantPtr p_value);
+typedef GDNativeBool (*GDNativeExtensionClassGet)(GDExtensionClassInstancePtr p_instance, const GDNativeStringNamePtr p_name, GDNativeVariantPtr r_ret);
+
+struct GDNativePropertyInfo {
+	uint32_t type;
+	const char *name;
+	const char *class_name;
+	uint32_t hint;
+	const char *hint_string;
+	uint32_t usage;
```

Comment:
I wonder if we should also add `PROPERTY_HINT` and `PROPERTY_USAGE` enums to use with this. I know we can get this from the JSON but it's not very convenient when using this with pure C.

Reply:
I would probably avoid adding anything that is in the JSON to keep the API as simple as possible. If you want to actually use the Godot API from pure C for some strange reason, we could do pure C bindings too ala GTK.

Reply:
we were also discussing with @groud about making the hints string only, although its a lot of work to change..